docs/ops/config.rst
Configuration
=============
All configuration happens through environment variables.

Django settings
---------------
These settings map directly to built-in Django settings. An environment
variable like ``DJANGO_FOO`` controls the Django setting ``FOO``. Not all
Django settings are available for configuration.

.. note::

    Default values given refer to ``Production`` configurations, which is the
    default in Docker images.. Other configurations may have defaults not
    listed here.

.. envvar:: DATABASE_URL

    :default: ``postgres://postgres@localhost/normandy``
    :documentation: https://github.com/kennethreitz/dj-database-URL#URL-schema

    A database URL, including any username and password, if needed.

.. envvar:: DJANGO_ALLOWED_HOSTS

    :default: Empty string
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:allowed-hosts

    A comma-seperated list of host values to accept. Examples:
    ``example.com,www.example.com``, ``.example.com``
    (matches any subdomain of example.com), ``*`` (allows everything). This
    setting is required to be set. If there are other protections (such as
    load balancers), setting this to ``*`` presents no risk.

.. envvar:: DJANGO_CONFIGURATION

    :default: ``Production``
    :documentation: http://django-configurations.readthedocs.io/en/stable/

    The name of a configuration preset to use for this environment. Useful
    values are

    * ``Production`` - Preferred
    * ``ProductionInsecure`` - For systems running without HTTPS
    * ``Build`` - This is used during CI to build static assets
    * ``Development`` - This is used by developers
    * ``Test`` - Used during unit tests

.. envvar:: DJANGO_DEBUG

    :default: ``false``
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:debug

    ``true`` or ``false``. Enables Django's debug mode. This should never be
    enabled on permanent servers. It is inefficient and leaks memory.

.. envvar:: DJANGO_MEDIA_URL

    :default: ``/media/``
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:media-url

    The URL prefix for media files (files uploaded to the service). Both
    host-relative (``/media/``) and host-absolute URLs
    (``https://cdn.example.com/``) work. Should end in a slash.

.. envvar:: DJANGO_SECRET_KEY

    :required: True
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:setting-SECRET_KEY

    A string used as a seed for security features. This should be the same
    on all instances that share a database, and should be kept secret. It
    should be a long, random string. This field is required to be set in
    most environments.

.. envvar:: DJANGO_STATIC_URL

    :default: ``/static/``
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:static-url

    The URL prefix for static files (files shipped with the service). Both
    host-relative and host-absolute URLs work. Should end in a slash.

.. envvar:: DJANGO_CONN_MAX_AGE

    :default: ``0``
    :documentation: https://docs.djangoproject.com/en/1.9/ref/settings/#std:setting-CONN_MAX_AGE

    Time to hold database connections open in seconds. If set to 0, will close
    every database connection immediately. Each worker (as controlled by
    ``WEB_CONCURRENCY``) will have its own connection.

Normandy settings
-----------------
These settings are specific to Normandy. In other words, they won't be present
in other Django projects.

.. envvar:: DJANGO_GEOIP2_DATABASE

    :default: ``/app/GeoIP2-Country.mmdb``

    Path to a Maxmind GeoIP Country database.

.. envvar:: DJANGO_ADMIN_ENABLED

    :default: ``true``

    For security, Normandy can disable write access. This should be enabled on
    production servers. Servers with this setting set to ``false`` shouldn't
    require write access to Postgres.

.. envvar:: DJANGO_IMMUTABLE_CACHE_TIME

    :default: ``31536000`` (1 year)

    Sets the time in seconds immutable objects (such as Action implementations)
    are cached for with the HTTP ``Cache-Control`` header.

.. envvar:: DJANGO_NUM_PROXIES

    :default: ``0``

    The number of proxies between users and Normandy. This is used to parse
    the ``X-Forwarded-For`` header.

.. envvar:: DJANGO_RAVEN_CONFIG_DSN

    :default: ``None``

    Optional. The DSN for Raven to report errors to Sentry.

.. envvar:: DJANGO_RAVEN_CONFIG_RELEASE

    :default: ``None``

    Optional. The release for Raven to report to Sentry. Automatically set by
    production Docker images.

.. envvar:: DJANGO_AUTOGRAPH_URL

    The URL where an Autograph_ server can be reached. If left blank, content
    signing will be disabled.

    .. _Autograph: https://github.com/mozilla-services/autograph/

.. envvar:: DJANGO_AUTOGRAPH_HAWK_ID

    The pre-arranged ID to use for Hawk authentication with Autograph.

.. envvar:: DJANGO_AUTOGRAPH_SECRET_KEY

    The pre-arranged secret key to use for Hawk authentication with Autograph.

.. envvar:: DJANGO_AUTOGRAPH_SIGNATURE_MAX_AGE

    :default: ``604800`` (1 week)

    Content with signature ages older than this are considered out of date and
    will be re-signed. The keys used by Autograph to sign content are generally
    only valid for a few weeks, and have a period of overlap where both the new
    key and old key are valid. The aim with this setting is to be as long as
    possible while still guaranteeing that actions will get resigned during the
    overlap period.

.. envvar:: DJANGO_X5U_CACHE_TIME

    :default: ``600`` (10 minutes)

    The time in seconds to cache the public keys retrieved from x5u URLs when
    verifying signatures. Set to 0 to disable caching.

.. envvar:: DJANGO_X5U_ERROR_CACHE_TIME

    :default: ``5``

    The time in seconds to cache errors while trying to retrieve public keys
    from x5u URLs when verifying signatures. Set to 0 to disable error caching.

.. envvar:: DJANGO_X5U_REQUEST_TIMEOUT

    :default: ``0.5``

    The time in seconds to wait to receive a response from the server when
    requesting x5u URLs to verify signatures. A value of 0 means no timeout.

.. envvar:: DJANGO_AUTOGRAPH_X5U_CACHE_BUST

    :default: Unset

    If set, the value will be added to the URL for the x5u certificate chain as
    a query parameter named `cachebust`. This is used to force clients to
    re-fetch the certificate chain in cases where they're caching an expired or
    otherwise invalid copy of the chain.

.. envvar:: DJANGO_REMOTE_SETTINGS_URL

    :default: Unset

    The URL where a `Remote Settings`_ server can be reached
    (e.g. ``https://settings-writer.prod.mozaws.net/v1``).

    If left blank, the publication of recipes will be disabled.

    .. _Remote Settings: https://remote-settings.readthedocs.io

.. envvar:: DJANGO_REMOTE_SETTINGS_USERNAME

    :default: Unset

    The account username that is allowed to create records on the
    Remote Settings collection.

.. envvar:: DJANGO_REMOTE_SETTINGS_PASSWORD

    :default: Unset

    The account password to authenticate as DJANGO_REMOTE_SETTINGS_USERNAME.

.. envvar:: DJANGO_REMOTE_SETTINGS_CAPABILITIES_COLLECTION_ID

    :default: ``normandy-recipes-capabilities``

    The name of the Remote Settings collection where the recipes will be published.

.. envvar:: DJANGO_REMOTE_SETTINGS_BUCKET_ID

    :default: ``main-workspace``

    The name of the Remote Settings bucket where the recipes will be published.

.. envvar:: REMOTE_SETTINGS_RETRY_REQUESTS

    :default: ``3``

    If the Remote Settings server does not return a successful response, the
    requests will be retried if the specified number is superior to zero.

.. envvar:: DJANGO_API_CACHE_TIME

    :default: ``30``

    The time in seconds to set in cache headers for cacheable APIs. This may be
    set to 0 in non-production environments to ease testing. In production
    environments, setting this value too low can be a denial-of-service risk.

.. envvar:: DJANGO_API_ROOT_CACHE_TIME

    :default: ``86400`` (1 day)

    The time in seconds to set in cache headers for cacheable API root
    endpoints.

.. envvar:: DJANGO_API_CACHE_ENABLED

    :default: ``True``

    Controls cache headers for cacheable APIs. If true, API views will send
    headers indicating that they can be cached according to
    :envvar:`DJANGO_API_CACHE_TIME`. If false, API views will send headers
    indicating that they should never be cached.

.. envvar:: DJANGO_PERMANENT_REDIRECT_CACHE_TIME

   :default: ``2592000`` (30 days)

   The time in seconds to set in cache headers for permanent redirects.

.. envvar:: DJANGO_HTTPS_REDIRECT_CACHE_TIME

   :default: ``2592000`` (30 days)

   The time in seconds to set in cache headers for permanent redirects
   to change from HTTP to HTTPS.

.. envvar:: DJANGO_LOGGING_USE_JSON

    :default: ``True``

    If this setting is true, standard logging will be output in mozlog_ format.
    Otherwise logs will be unstructured.

    .. _mozlog: https://github.com/mozilla-services/Dockerflow/blob/master/docs/mozlog.md

.. envvar:: DJANGO_CSP_REPORT_URI

   :default: ``None``

   Controls the ``report-uri`` directive in the Content Security Policy header.
   Attempts to violate the Content Security Policy are sent by the browser to
   this URL. See the `MDN documentation on report-uri <report-uri>`_ for more
   info.

   .. _report-uri: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri

.. envvar:: DJANGO_CDN_URL

   :default: ``None``

   The URL of a CDN that is backed by Normandy, if one is in use. This is used to
   enforce that immutable content is routed through the CDN. Must end with a
   slash (``/``).

.. envvar:: DJANGO_APP_SERVER_URL

   :default: ``None``

   The URL that allows direct access to Normandy, bypassing any CDNs. This
   is used for content that cannot be cached. If not specified, Normandy will
   assume direct access. Must end with a slash (``/``).

.. envvar:: DJANGO_USE_OIDC

   :default: ``False``

   If enabled, Normandy will authenticate users by reading a header in requests.
   The expectation is that a proxy server, such as Nginx, will perform
   authentication using Open ID Connect, and then pass the unique ID of the user
   in a header.

   .. seealso::

      :envvar:`DJANGO_OIDC_REMOTE_AUTH_HEADER` for which header Normandy
      reads this value from.

   .. warning::

      If this feature is enabled, the proxy server providing authentication
      *must* sanitize the headers passed along to Normandy. Specifically, the
      header defined in :envvar:`DJANGO_OIDC_REMOTE_AUTH_HEADER` must not be
      passed on from the user.

      Failing to do this will result in any client being able to authenticate
      as any user, with no checks.

.. envvar:: DJANGO_OIDC_REMOTE_AUTH_HEADER

   :default: ``HTTP_REMOTE_USER``

   If :envvar:`DJANGO_USE_OIDC` is ``True``, this is the source of the user to
   authenticate. This must match Django header normalization, i.e. it must be
   capitalized, dashes replaced with underscores, and be prefixed with ``HTTP_``.

   For example, the header ``OIDC-Claim-User-Profile-Email`` becomes
   ``HTTP_OIDC_CLAIM_USER_PROFILE_EMAIL``.

.. envvar:: DJANGO_OIDC_LOGOUT_URL

   If :envvar:`DJANGO_USE_OIDC` is set to ``True``, this settings must be set to
   the URL that a user can visit to logout. It may be a relative URL.

.. envvar:: DJANGO_PEER_APPROVAL_ENFORCED

   :default: ``True``

   If ``True``, approval requests for recipe changes can only be approved by a
   different user than the one who created the request. If ``False``, approval
   requests can be approved by the same user who created it.

   This defaults to ``False`` for local developer instances.

.. envvar:: DJANGO_CERTIFICATES_CHECK_VALIDITY

   :default: ``True``

   Whether to check that certificates used for recipe signatures are checked
   for validity. This means that the date ranges in the certificate are
   checked against the current date. This should likely never be turned off
   in Production.

.. envvar:: DJANGO_CERTIFICATES_EXPIRE_EARLY_DAYS

   :default: ``None``

   If set, when checking certificates for validity, start failing
   system checks this many days before the certificate would expire.

.. envvar:: DJANGO_CERTIFICATES_EXPECTED_ROOT_HASH

   :default: ``None``

   If this is set to a string, certificates will be checked to be originating
   from a trusted source, by hashing the root certificate in the x5u
   certificate chain. For possible values of this setting, see the :ref:`list of
   environments <recipes>`.


normandy/settings.py
import json
import os

from configurations import Configuration, values
from corsheaders.defaults import default_methods


class Core(Configuration):
    """Settings that will never change per-environment."""

    # Build paths inside the project like this: os.path.join(BASE_DIR, ...)
    BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

    # Application definition
    INSTALLED_APPS = [
        "normandy.base.apps.BaseApp",
        "normandy.health.apps.HealthApp",
        "normandy.recipes.apps.RecipesApp",
        "normandy.selfrepair",
        "normandy.studies",
        "product_details",
        "rest_framework",
        "graphene_django",
        "drf_yasg2",
        "storages",
        "raven.contrib.django.raven_compat",
        "corsheaders",
        "django_filters",
        "dockerflow.django",
        "django.contrib.auth",
        "django.contrib.contenttypes",
        "django.contrib.sessions",
        "django.contrib.messages",
        "whitenoise.runserver_nostatic",
        "django.contrib.staticfiles",
    ]

    # Middleware that ALL environments must have. See the Base class for
    # details.
    MIDDLEWARE = [
        "normandy.base.middleware.response_metrics_middleware",
        "corsheaders.middleware.CorsMiddleware",
        "normandy.base.middleware.request_received_at_middleware",
        "dockerflow.django.middleware.DockerflowMiddleware",
        "normandy.base.middleware.NormandySecurityMiddleware",
        "normandy.base.middleware.NormandyWhiteNoiseMiddleware",
        "normandy.base.middleware.NormandyCommonMiddleware",
        "normandy.base.middleware.show_yaml_in_browser_middleware",
        "django.middleware.clickjacking.XFrameOptionsMiddleware",
        "csp.middleware.CSPMiddleware",
    ]

    ROOT_URLCONF = "normandy.urls"

    TEMPLATES = [
        {
            "BACKEND": "django.template.backends.django.DjangoTemplates",
            "DIRS": [],
            "APP_DIRS": True,
            "OPTIONS": {
                "context_processors": [
                    "django.template.context_processors.debug",
                    "django.template.context_processors.request",
                    "django.contrib.auth.context_processors.auth",
                    "django.contrib.messages.context_processors.messages",
                    "django.template.context_processors.static",
                ]
            },
        }
    ]

    WSGI_APPLICATION = "normandy.wsgi.application"

    # Internationalization
    LANGUAGE_CODE = "en-us"
    TIME_ZONE = "UTC"
    USE_I18N = False
    USE_L10N = False
    USE_TZ = True

    # Static files (CSS, JavaScript, Images)
    STATICFILES_FINDERS = [
        "django.contrib.staticfiles.finders.FileSystemFinder",
        "django.contrib.staticfiles.finders.AppDirectoriesFinder",
        "npm.finders.NpmFinder",
    ]

    REST_FRAMEWORK = {
        "DEFAULT_AUTHENTICATION_CLASSES": [
            "normandy.base.api.authentication.BearerTokenAuthentication"
        ],
        "DEFAULT_FILTER_BACKENDS": ["django_filters.rest_framework.DjangoFilterBackend"],
        "TEST_REQUEST_DEFAULT_FORMAT": "json",
        "DEFAULT_RENDERER_CLASSES": (
            "normandy.base.api.renderers.CanonicalJSONRenderer",
            "normandy.base.api.renderers.CustomBrowsableAPIRenderer",
            "rest_framework_yaml.renderers.YAMLRenderer",
        ),
        "DEFAULT_PAGINATION_CLASS": "rest_framework.pagination.PageNumberPagination",
        "PAGE_SIZE": 25,
        "EXCEPTION_HANDLER": "normandy.base.api.views.exception_handler",
        "DEFAULT_VERSIONING_CLASS": "rest_framework.versioning.NamespaceVersioning",
    }

    GRAPHENE = {"SCHEMA": "normandy.schema.schema"}

    # Content Security Policy
    def CSP_DEFAULT_SRC(self):
        srcs = ["'self'"]
        if self.CDN_URL:
            srcs.append(self.CDN_URL)
        return srcs

    CSP_OBJECT_SRC = "'none'"  # not using <object>, <embed>, and <applet> elements
    CSP_WORKER_SRC = "'none'"  # not using JS Worker, SharedWorker, or ServiceWorkers
    CSP_FRAME_SRC = "'none'"  # not using frames or iframes
    CSP_BLOCK_ALL_MIXED_CONTENT = True

    # these don't fallback to default-src
    CSP_BASE_URI = "'none'"  # not using <base>
    CSP_FORM_ACTION = "'self'"  # we only submit forms to ourselves

    # TODO(mythmon) Re-add this once either:
    #   1) We know the domain we can use for Firefox's framing of the self-repair frame
    #   2) We've deprecated the self-repair frame entirely
    #   3) django-csp allows removing directives on an individual view (mozilla/django-csp#85)
    # CSP_FRAME_ANCESTORS = "'none'"  # Block framing by default

    # Directory where we, locally, have a list of directories of actions' implementations.
    ACTIONS_ROOT_DIRECTORY = os.path.join(BASE_DIR, "client/actions")

    # Actions, when found on disk, whose name needs to be converted before saved in the ORM.
    ACTIONS_ALIAS_NAMES = {"single-preference-experiment": "preference-experiment"}

    # Path on disk where the @mozilla/normandy-action-argument-schemas is installed.
    # This node package is expected to have been installed as a pre-build step
    # (i.e. `yarn`) and should always exist.
    ACTIONS_SCHEMA_DIRECTORY = values.Value(
        os.path.join(BASE_DIR, "node_modules", "@mozilla", "normandy-action-argument-schemas")
    )

    PROD_DETAILS_STORAGE = values.Value("normandy.recipes.storage.ProductDetailsRelationalStorage")

    AWS_QUERYSTRING_AUTH = False

    # We changed the CSRF cookie from http-only to non http-only and need to override existing
    # cookies. The easiest way is just change the cookie name when such changes happen.
    CSRF_COOKIE_NAME = "csrftoken-20170707"


# ==================== MIXINS ====================
# These mixins are used to group settings together, and then are included in the environments below


class CORS:
    """Default settings related to setting CORS headers."""

    CORS_ORIGIN_ALLOW_ALL = values.BooleanValue(True)
    CORS_URLS_REGEX = r"^/api/.*$"
    CORS_ALLOW_METHODS = values.ListValue(default_methods)
    # This list only applicable if you've also set
    # `DJANGO_CORS_ORIGIN_ALLOW_ALL` to False.
    CORS_ORIGIN_WHITELIST = values.ListValue([])


class OIDC:
    """Settings related to talking to an OIDC provider for authorizing
    access tokens received from the client."""

    # Note! We *could* just have a setting called 'OIDC_DOMAIN' and then use
    # https://$OIDC_DOMAIN/.well-known/openid-configuration and extract the
    # 'userinfo_endpoint' value from that during startup or something.
    # As of May 2018, the likelyhood of this URL changing and the fact that it's
    # the only URL we need, let's just make the setting the URL that we need
    # for being able to authorization by access token.
    OIDC_USER_ENDPOINT = values.URLValue("https://auth.mozilla.auth0.com/userinfo")


class InsecureAuthentication:
    def REST_FRAMEWORK(self):
        return {
            **Core.REST_FRAMEWORK,
            "DEFAULT_AUTHENTICATION_CLASSES": Core.REST_FRAMEWORK["DEFAULT_AUTHENTICATION_CLASSES"]
            + ["normandy.base.api.authentication.InsecureEmailAuthentication"],
        }


class Metrics:
    METRICS_USE_DEBUG_LOGS = values.BooleanValue(False)
    METRICS_USE_STATSD = values.BooleanValue(False)
    METRICS_STATSD_HOST = values.Value("localhost")
    METRICS_STATSD_PORT = values.IntegerValue(8125)
    METRICS_STATSD_NAMESPACE = values.Value("")


# ==================== ENVIRONMENTS ====================


class Base(Core, CORS, OIDC, Metrics):
    """Settings that may change per-environment, some with defaults."""

    # Flags that affect other settings, via setting methods below
    LOGGING_USE_JSON = values.BooleanValue(False)
    USE_OIDC = values.BooleanValue(False)

    # General settings
    DEBUG = values.BooleanValue(False)
    ADMINS = values.SingleNestedListValue([])
    SILENCED_SYSTEM_CHECKS = values.ListValue(
        [
            # We've subclassed Django's security middleware, so Django's
            # checks can't tell we are using the middleware.
            "security.W001",
            # We have disabled CSRF as we do not use cookie-based authentication.
            # As a result the CSRF middleware check is not required.
            "security.W003",
        ]
    )

    # Authentication
    def AUTHENTICATION_BACKENDS(self):
        if self.USE_OIDC:
            return ["normandy.base.auth_backends.EmailOnlyRemoteUserBackend"]
        else:
            return ["normandy.base.auth_backends.LoggingModelBackend"]

    OIDC_REMOTE_AUTH_HEADER = values.Value("HTTP_REMOTE_USER")

    # Middleware that _most_ environments will need. Subclasses can override this list.
    EXTRA_MIDDLEWARE = [
        "django.contrib.sessions.middleware.SessionMiddleware",
        "django.contrib.auth.middleware.AuthenticationMiddleware",
        "django.contrib.messages.middleware.MessageMiddleware",
    ]

    def MIDDLEWARE(self):
        """
        Determine middleware by combining the core set and
        per-environment set.
        """
        middleware = Core.MIDDLEWARE + self.EXTRA_MIDDLEWARE
        if self.USE_OIDC:
            middleware.append("normandy.base.middleware.ConfigurableRemoteUserMiddleware")
        return middleware

    def LOGGING(self):
        return {
            "version": 1,
            "disable_existing_loggers": False,
            "formatters": {
                "json": {"()": "dockerflow.logging.JsonLogFormatter", "logger_name": "normandy"},
                "development": {"format": "%(levelname)s %(asctime)s %(name)s %(message)s"},
            },
            "handlers": {
                "console": {
                    "level": "DEBUG",
                    "class": "logging.StreamHandler",
                    "formatter": "json" if self.LOGGING_USE_JSON else "development",
                }
            },
            "root": {"handlers": ["console"], "level": "WARNING"},
            "loggers": {
                "normandy": {"propagate": False, "handlers": ["console"], "level": "DEBUG"},
                "request.summary": {"propagate": False, "handlers": ["console"], "level": "DEBUG"},
            },
        }

    # Remote services
    DATABASES = values.DatabaseURLValue("postgres://postgres@localhost/normandy")
    CONN_MAX_AGE = values.IntegerValue(0)
    GEOIP2_DATABASE = values.Value(os.path.join(Core.BASE_DIR, "GeoLite2-Country.mmdb"))
    # Email settings
    EMAIL_HOST_USER = values.Value()
    EMAIL_HOST = values.Value()
    EMAIL_PORT = values.IntegerValue(587)
    EMAIL_USE_TLS = values.BooleanValue(True)
    EMAIL_HOST_PASSWORD = values.Value()
    EMAIL_BACKEND = values.Value("django.core.mail.backends.smtp.EmailBackend")

    def RAVEN_CONFIG(self):
        version_path = os.path.join(Core.BASE_DIR, "version.json")
        try:
            with open(version_path) as f:
                build_info = json.loads(f.read())
                version = build_info.get("version")
        except IOError:
            version = None

        return {
            "dsn": values.URLValue(None, environ_name="RAVEN_CONFIG_DSN"),
            "release": values.Value(version, environ_name="RAVEN_CONFIG_RELEASE"),
        }

    # statsd
    STATSD_HOST = values.Value("localhost")
    STATSD_PORT = values.IntegerValue(8125)
    STATSD_IPV6 = values.BooleanValue(False)
    STATSD_PREFIX = values.Value("normandy")
    STATSD_MAXUDPSIZE = values.IntegerValue(512)

    # Security settings
    SECRET_KEY = values.SecretValue()
    ALLOWED_HOSTS = values.ListValue([])
    AUTH_PASSWORD_VALIDATORS = [
        {"NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator"},
        {"NAME": "django.contrib.auth.password_validation.MinimumLengthValidator"},
        {"NAME": "django.contrib.auth.password_validation.CommonPasswordValidator"},
        {"NAME": "django.contrib.auth.password_validation.NumericPasswordValidator"},
    ]
    PASSWORD_HASHERS = values.ListValue(
        [
            "django.contrib.auth.hashers.BCryptSHA256PasswordHasher",
            "django.contrib.auth.hashers.PBKDF2PasswordHasher",
        ]
    )
    USE_X_FORWARDED_HOST = values.BooleanValue(False)
    SECURE_PROXY_SSL_HEADER = values.TupleValue()
    SECURE_HSTS_SECONDS = values.IntegerValue(3600)
    SECURE_HSTS_INCLUDE_SUBDOMAINS = values.BooleanValue(True)
    CSRF_COOKIE_HTTPONLY = values.BooleanValue(True)
    CSRF_COOKIE_SECURE = values.BooleanValue(True)
    SECURE_SSL_REDIRECT = values.BooleanValue(True)
    SECURE_REDIRECT_EXEMPT = values.ListValue([])
    SESSION_COOKIE_SECURE = values.BooleanValue(True)
    SECURE_BROWSER_XSS_FILTER = values.BooleanValue(True)
    SECURE_CONTENT_TYPE_NOSNIFF = values.BooleanValue(True)
    X_FRAME_OPTIONS = values.Value("DENY")
    REQUIRE_RECIPE_AUTH = values.BooleanValue(True)

    # Media and static settings
    STATIC_URL = values.Value("/static/")
    STATIC_ROOT = values.Value(os.path.join(Core.BASE_DIR, "static"))
    MEDIA_URL = values.Value("/media/")
    MEDIA_ROOT = values.Value(os.path.join(Core.BASE_DIR, "media"))

    STATICFILES_DIRS = (os.path.join(Core.BASE_DIR, "assets"),)

    # URL that the CDN exists at to front cached parts of the site, if any.
    CDN_URL = values.URLValue(None)
    # URL that bypasses any CDNs
    APP_SERVER_URL = values.URLValue(None)

    # URL for the CSP report-uri directive.
    CSP_REPORT_URI = values.Value("/__cspreport__")

    # Normandy settings
    ADMIN_ENABLED = values.BooleanValue(True)
    IMMUTABLE_CACHE_TIME = values.IntegerValue(60 * 60 * 24 * 365)
    NUM_PROXIES = values.IntegerValue(0)
    API_ROOT_CACHE_TIME = values.IntegerValue(60 * 60 * 24)
    API_CACHE_TIME = values.IntegerValue(30)
    API_CACHE_ENABLED = values.BooleanValue(True)
    PERMANENT_REDIRECT_CACHE_TIME = values.IntegerValue(60 * 60 * 24 * 30)
    HTTPS_REDIRECT_CACHE_TIME = values.IntegerValue(60 * 60 * 24 * 30)
    X5U_CACHE_TIME = values.IntegerValue(60 * 10)
    X5U_ERROR_CACHE_TIME = values.IntegerValue(5)
    X5U_REQUEST_TIMEOUT = values.IntegerValue(0.5)

    # If true, approvals must come from two separate users. If false, the same
    # user can approve their own request.
    PEER_APPROVAL_ENFORCED = values.BooleanValue(True)

    # Autograph settings
    AUTOGRAPH_URL = values.Value()
    AUTOGRAPH_HAWK_ID = values.Value()
    AUTOGRAPH_HAWK_SECRET_KEY = values.Value()
    AUTOGRAPH_SIGNATURE_MAX_AGE = values.IntegerValue(60 * 60 * 24 * 7)
    AUTOGRAPH_X5U_CACHE_BUST = values.Value(None)

    # Remote Settings connection configuration
    REMOTE_SETTINGS_URL = values.Value()
    REMOTE_SETTINGS_USERNAME = values.Value()
    REMOTE_SETTINGS_PASSWORD = values.Value()
    REMOTE_SETTINGS_PUBLISH_BUCKET_ID = values.Value("main")
    REMOTE_SETTINGS_WORKSPACE_BUCKET_ID = values.Value("main-workspace")
    REMOTE_SETTINGS_CAPABILITIES_COLLECTION_ID = values.Value("normandy-recipes-capabilities")
    REMOTE_SETTINGS_RETRY_REQUESTS = values.IntegerValue(3)

    # How many days before expiration to warn for expired certificates
    CERTIFICATES_EXPIRE_EARLY_DAYS = values.IntegerValue(None)
    CERTIFICATES_CHECK_VALIDITY = values.BooleanValue(True)
    CERTIFICATES_EXPECTED_ROOT_HASH = values.Value(None)
    CERTIFICATES_EXPECTED_SUBJECT_CN = values.Value("normandy.content-signature.mozilla.org")

    # Storage settings
    DEFAULT_FILE_STORAGE = values.Value("normandy.base.storage.NormandyS3Boto3Storage")

    PROD_DETAILS_DIR = values.Value(os.path.join(Core.BASE_DIR, "product_details"))

    # AWS settings
    AWS_ACCESS_KEY_ID = values.Value()
    AWS_SECRET_ACCESS_KEY = values.Value()
    AWS_STORAGE_BUCKET_NAME = values.Value()
    GS_BUCKET_NAME = values.Value()
    GS_DEFAULT_ACL = values.Value("publicRead")

    GITHUB_URL = values.Value("https://github.com/mozilla/normandy")

    BASELINE_CAPABILITIES = values.SetValue(
        set(
            [
                "action.addon-rollback",
                "action.addon-rollout",
                "action.addon-study",
                "action.branched-addon-study",
                "action.console-log",
                "action.multi-preference-experiment",
                "action.opt-out-study",
                "action.preference-experiment",
                "action.preference-rollback",
                "action.preference-rollout",
                "action.show-heartbeat",
                "action.single-preference-experiment",
                "jexl.transform.bucketSample",
                "jexl.transform.date",
                "jexl.transform.keys",
                "jexl.transform.preferenceExists",
                "jexl.transform.preferenceIsUserSet",
                "jexl.transform.preferenceValue",
                "jexl.transform.stableSample",
            ]
        )
    )


class Development(InsecureAuthentication, Base):
    """Settings for local development."""

    DOTENV_EXISTS = os.path.exists(os.path.join(Core.BASE_DIR, ".env"))
    DOTENV = ".env" if DOTENV_EXISTS else None

    SECRET_KEY = values.Value("not a secret")
    DEBUG = values.BooleanValue(True)
    AUTH_PASSWORD_VALIDATORS = values.ListValue([])
    INSTALLED_APPS = Base.INSTALLED_APPS + ["django_extensions"]
    EMAIL_BACKEND = values.Value("django.core.mail.backends.console.EmailBackend")
    SECURE_SSL_REDIRECT = values.Value(False)
    REQUIRE_RECIPE_AUTH = values.BooleanValue(False)
    PEER_APPROVAL_ENFORCED = values.BooleanValue(False)
    CSP_REPORT_URI = values.Value("")
    DEFAULT_FILE_STORAGE = values.Value("django.core.files.storage.FileSystemStorage")

    API_CACHE_ENABLED = values.BooleanValue(False)
    API_ROOT_CACHE_TIME = values.IntegerValue(0)
    API_CACHE_TIME = values.IntegerValue(0)

    SILENCED_SYSTEM_CHECKS = values.ListValue(["normandy.recipes.E006"])  # geoip db not available
    METRICS_USE_DEBUG_LOGS = values.Value(True)

    def LOGGING(self):
        config = super().LOGGING()
        config["loggers"]["werkzeug"] = {
            "handlers": ["console"],
            "level": "DEBUG",
            "propagate": False,
        }
        config["loggers"]["markus"] = {
            "handlers": ["console"],
            "level": "INFO",
            "propagate": False,
        }
        return config


class Production(Base):
    """Settings for the production environment."""

    SECURE_PROXY_SSL_HEADER = values.TupleValue(("HTTP_X_FORWARDED_PROTO", "https"))
    LOGGING_USE_JSON = values.Value(True)
    SECURE_HSTS_SECONDS = values.IntegerValue(31536000)  # 1 year
    DEFAULT_FILE_STORAGE = values.Value("normandy.base.storage.NormandyS3Boto3Storage")
    AWS_S3_FILE_OVERWRITE = False
    METRICS_USE_STATSD = values.Value(True)


class ProductionReadOnly(Production):
    """
    Settings for a production environment that is read-only. This is
    used on public-facing webheads.
    """

    EXTRA_MIDDLEWARE = [
        # No need for sessions, so removing those middlewares helps us go fast
    ]
    ADMIN_ENABLED = values.BooleanValue(False)
    SILENCED_SYSTEM_CHECKS = values.ListValue(
        ["security.W001", "security.W003"]  # Security middle ware check  # CSRF middleware check
    )

    # In ReadOnly mode you have no business using the API to do any writes
    # anyway. This gives early feedback to that.
    CORS_ALLOW_METHODS = values.ListValue(["GET", "OPTIONS"])


class ProductionInsecure(Production):
    """
    Settings for a production-like environment that lacks many security features.

    Useful for testing and setups where security is provided by other means.
    Not intended for general use on the public internet.
    """

    SECRET_KEY = values.Value("not a secret")
    ALLOWED_HOSTS = values.ListValue(["*"])
    SECURE_SSL_REDIRECT = values.BooleanValue(False)
    CSRF_COOKIE_SECURE = values.BooleanValue(False)
    SECURE_HSTS_SECONDS = values.IntegerValue(0)
    SESSION_COOKIE_SECURE = values.BooleanValue(False)

    # These checks aren't useful for a purposefully insecure environment
    SILENCED_SYSTEM_CHECKS = values.ListValue(
        [
            "security.W001",  # security middleware check
            "security.W003",  # CSRF middleware check
            "security.W004",  # check hsts seconds
            "security.W008",  # Secure SSL redirect
            "security.W009",  # Secret key length
            "security.W012",  # Check session cookie secure
            "security.W016",  # Check CSRF cookie secure
        ]
    )


class ProductionReadOnlyInsecure(ProductionInsecure, ProductionReadOnly):
    pass


class Build(Production):
    """Settings for building the Docker image for production."""

    SECRET_KEY = values.Value("not a secret")


class Test(InsecureAuthentication, Base):
    DOTENV_EXISTS = os.path.exists(os.path.join(Core.BASE_DIR, ".env"))
    DOTENV = ".env" if DOTENV_EXISTS else None
    SECRET_KEY = "not a secret"
    # Don't allow tests use any storage. Use the `storage` fixture for tests
    # that need access to storage
    DEFAULT_FILE_STORAGE = "normandy.base.storage.NotAllowedStorage"
    SECURE_SSL_REDIRECT = False
    AUTOGRAPH_URL = None
    AUTOGRAPH_HAWK_ID = None
    AUTOGRAPH_HAWK_SECRET_KEY = None
    OIDC_USER_ENDPOINT = "https://auth.example.com/userinfo"


class Docs(Base):
    SECRET_KEY = "not a secret"
